docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+**This roadmap is** a sketch, rather than a final design for the “forward” (using C++ APIs from Swift) half of C++ and Swift interoperability. This roadmap outlines some high-level topics related to C++ interoperability, overarching goals that drive the project’s design decisions, a process for evolving C++ interoperability over time, and a collection of specific problems and their potential solutions. 
+
+“Reverse” interoperability (using Swift APIs from C++) is another extremely important part of the interoperability story, however, reverse interoperability has largely different goals and constraints, which necessarily mean a different design and roadmap. Alex Lorenz will be putting together a roadmap for reverse interoperability which outlines these goals and will shed light onto the differences between the two halves of the projects. In the future this roadmap will need to be updated to explain how these two halves fit together and present the “bi-directional” interoperability story that is C++ and Swift interoperability as a whole. 
+
+**C++, as a language:** C++ is an stubbornly un-opinionated language. There is little to no consensus on anything from naming, to generic programming, value categories, or error handling (to name a few things). This lack of opinion is in sharp contrast to Swift where there is generally one, specific way to do things. Even in the C++ Standard Library there is a rather un-standard set of API models. The un-opinionated nature of C++ means there are few, if any guarantees about a given C++ API, because the same functionality may be implemented using many different language features or programming paradigms. When consuming an arbitrary C++ API, you must be prepared for *anything*. When consuming Swift APIs, this is not the case. Swift is an opinionated language which means rather than consuming *any* API pattern you are consuming *some*, specific API pattern. Swift clearly defines patterns for naming, generic programming, value categories, error handling, etc. and it is rare that a codebase deviates from the standard practice. 
+
+**C and Objective-C interoperability:** C++ interoperability will have to build off of the previous art of C and Objective-C interoperability. Starting with the name, C++ is a superset of C, which means C++ interoperability has a great starting place to build off of. C is a simple language without many high level constructs, so importing C APIs is, relatively, simple. On the other hand, Objective-C adds some high level constructs, such as reference types and a generics model. However, unlike C++, Objective-C has strong idioms; there are only so many patterns for doing some specific thing. Further, Objective-C was able to be updated to better interoperate with Swift, the generics model is a good example of this. So, while C and Objective-C interoperability is a good starting place, the model for importing these languages cannot be taken wholesale for C++ interoperability. 
```

Comment:
And a good base line to compare our designs with.

Reply:
I am not sure I 100% follow the point about ObjC language being able to to be updated. In theory you could continue evolving C++. The issue is that interop needs to work with existing codebases, which was the issue for ObjC as well… This sentence could probably be just dropped.

Reply:
That's a good point. I'll drop it. 

I think there is something here about Objective-C being able to be evolved so it could interop with Swift, where I doubt the C++ committee would ever evolve C++ to make it interop better with Swift, but I don't think it's a necessary thing to call out here.

Reply:
I think your point is that ObjC and Swift are converging in some ways to make their interop easier, however that doesn't really help for the purposes of this roadmap, especially since your next sentence blows up any hope that we can use Swift/ObjC interop as a basis. As you said, I think it's best to drop.